pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"io/ioutil"
+	"math"
+	"net/http"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	mgr  MetricsManager
+	once sync.Once
+)
+
+const (
+	httpPattern = "/metrics"
+	addr        = ":12120"
+	httpAddr    = "http://localhost:12120/metrics"
+)
+
+func initMgr() {
+	mgr = NewMetricsManager()
+	if err := mgr.StartServing(httpPattern, addr, nil); err != nil {
+	}
+}
+
+func TestNew(t *testing.T) {
+	once.Do(initMgr)
+	if mgr == nil {
+		t.Errorf("failed testing new")
+	}
+}
+
+func TestStartServing(t *testing.T) {
+	once.Do(initMgr)
+	if _, err := http.Get(httpAddr); err != nil {
+		t.Errorf("failed to get response from server %v", err)
+	}
+}
+
+func TestOperationStart(t *testing.T) {
+	once.Do(initMgr)
+	// add an operation
+	op := Operation{
+		Name:       "op1",
+		Driver:     "driver1",
+		ResourceID: types.UID("uid1"),
+	}
+	mgr.OperationStart(op)
+	// should create an InterimFailure data point with latency ~ 300ms
+	time.Sleep(300 * time.Millisecond)
+	mgr.RecordMetrics(op, InterimFailure)
+	// should create an InterimFailure data point with latency ~ 1100ms
+	time.Sleep(800 * time.Millisecond)
+	mgr.RecordMetrics(op, Success)
```

Comment:
Comment says "InterimFailure" but it records a "Success".

Reply:
done